[Nicholas Tang Boon Keat] iP

NicholasT1 opened the pull request:
# YapBot
_A helpful assistant to organise your tasks_
> "10/10 Love it!" - Bob

Highlights:
 * Text-Based 
 * Easy to use
 * Ultra-low latency

Features:
- [x] Create Task
- [x] Create Task with deadline
- [x] Manage Tasks
- [x] Save Tasks
- [ ] Edit Tasks

---

How to Use:
1. Download jar file [here](https://github.com/NicholasT1/ip/releases)
2. Open a terminal and change to directory containing the file. 
3. Run `java -jar yapbot.jar`

Configure:

Tasks are saved by default to `data/tasks.txt`. You can change this in `YapBot.java`.
```java

public static void main(String[] args) {
    new YapBot("data/tasks.txt").run();
}

```
---
Hope this helped you! 🙃

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

NicholasT1 committed:
Renamed Duke to YapBot and changed greeting message

NicholasT1 committed:
Added echoing of user input

NicholasT1 committed:
Add list for storing user input

NicholasT1 committed:
Added Task class and changed list to store an array of Task type

NicholasT1 committed:
Add mark and unmark done feature

NicholasT1 committed:
Added Event, todo and deadline tasks

NicholasT1 committed:
Added automated testing

NicholasT1 committed:
Added exceptions and error handling in YapBot, updated test cases accordingly

NicholasT1 committed:
Add delete feature and add test cases for delete

NicholasT1 committed:
Added trim to input

NicholasT1 committed:
Add javadoc comments and adjust code style

NicholasT1 committed:
Add save and load features

NicholasT1 committed:
Merge branch 'branch-Level-7'

NicholasT1 committed:
Use LocalDateTime for deadlines

Deadlines are stored as plain strings.

Validation cannot be done to ensure date/times are valid.

NicholasT1 committed:
Add support for LocalDateTime in Event tasks

NicholasT1 committed:
Merge branch 'branch-Level-8'

Deadline fields are stored as String.

Deadline fields input should be validated to ensure valid dates and
times are being stored.

Let's use LocalDateTime since the user might want to set both dates and
times for their tasks.

NicholasT1 committed:
Move validation from YapBot to respective task constructors

NicholasT1 committed:
Add new classes for further abstraction. In progress.

NicholasT1 committed:
Extracted out relevant classes

jiayr99 commented on src/main/java/yapbot/tasks/ToDo.java:
```diff
@@ -0,0 +1,25 @@
+ * Child class of Task with no additional fields.
+ */
+public class ToDo extends Task {
+
```
some javadoc for the constructors could make the code easier to read?

jiayr99 commented on src/main/java/yapbot/exceptions/YapBotException.java:
```diff
@@ -0,0 +1,19 @@
+public class YapBotException extends Exception {
+    private static final String PREFIX_LINE = "\n-------------------------------------------\n";
+    private static final String POSTFIX_LINE = "\n-------------------------------------------\n";
+
```
I like that you added the lines to make the exception more prominent when printed out :+1:

jiayr99 commented on src/main/java/yapbot/main/YapBot.java:
```diff
@@ -0,0 +1,566 @@
+ */
+public class YapBot {
+    private static final String PREFIX_LINE = "\n-------------------------------------------";
+    private static final String POSTFIX_LINE = "-------------------------------------------\n";
```
the prefix and postfix line could maybe defined in another interface since it is used often?

jiayr99 commented on src/main/java/yapbot/main/YapBot.java:
```diff
@@ -0,0 +1,566 @@
+     * @param taskDetails Details of the task to be created.
+     * @throws YapBotException If task details are missing for the specified task type.
+     */
+    private static void createTask(Tasktype tasktype, String taskDetails) throws YapBotException {
```
I think this createTask method is really long, maybe it could be broken down?

jiayr99 reviewed (commented):
Overall i think your code is well done and well structured. I think the only issue is that certain methods tend to be very long, but other than that great job!!

NicholasT1 committed:
Add datetimeformatter constants with locale set to en_sg

NicholasT1 committed:
remove extra log message

NicholasT1 committed:
Move error message to display after welcome message

NicholasT1 committed:
Merge branch 'A-MoreOOP'

Commands and various functions are not encapsulated in their own
classes.

These functions such as Commands, tasklist, etc should be abstracted
into their own classes to increase readability and extensibility.

Let's:
 - Encapsulate tasks management into a TaskList class
 - Encapsulate user interaction and parsing of user input into Ui and
   Parser classes respectively
 - Encapsulate all commands into subclasses of an abstract Command class

NicholasT1 committed:
Merge branch 'add-gradle-support'

NicholasT1 committed:
Update mainclass in build.gradle file

NicholasT1 committed:
Add junit tests and further abstracted task creation

Unit tests are not implemented and creation of new tasks is still under
the TaskList class.

Lack of tests prevents meaningful testing of individual functions and
the lack of abstraction makes future additions more difficult.

Let's:
 - Add in Junit tests
 - Abstract away task creation into the respective commands

NicholasT1 committed:
Change filename of produced jar file

NicholasT1 committed:
Adjust code to fit coding standard

NicholasT1 committed:
Add JavaDoc comments for classes and methods

NicholasT1 committed:
Add find function

NicholasT1 committed:
Merge branch 'A-JavaDoc'

NicholasT1 committed:
Merge branch 'Level-9'

NicholasT1 committed:
Add javadoc comments for new methods

NicholasT1 committed:
Change isEmpty method call in getMatchingTasks to use length instead

SamsonChew commented on src/main/java/yapbot/util/Parser.java:
```diff
@@ -0,0 +1,74 @@
+package yapbot.util;
+
+import yapbot.commands.*;
```
Should this package details be listed out explicitly? Using a wildcard import may lead to naming conflicts sometime.

NicholasT1 committed:
Add checkstyle and adjust some style violations

NicholasT1 committed:
Add config files for checkstyle

NicholasT1 committed:
Merge branch 'A-CheckStyle'

SamsonChew commented on src/main/java/yapbot/tasks/Deadline.java:
```diff
@@ -0,0 +1,74 @@
+        LocalDateTime deadline;
+
+        // Converts string to LocalDateTime
+        if (deadlineStr.contains("AM") | deadlineStr.contains("PM")) {
```
Should this be || if you are using OR operator for this condition?

SamsonChew commented on src/main/java/yapbot/util/Storage.java:
```diff
@@ -0,0 +1,158 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+import java.util.Scanner;
```
I like how you organize the imports and make it consistent.

SamsonChew commented on src/main/java/yapbot/util/Storage.java:
```diff
@@ -0,0 +1,158 @@
+     * @throws YapBotException If the file cannot be parsed or the file cannot be created.
+     */
+    public ArrayList<Task> load() throws YapBotException {
+        try {
```
I like how you handle the errors here but perhaps can make the code more organized by creating another method to control the error handling?

SamsonChew commented on src/main/java/yapbot/tasks/Task.java:
```diff
@@ -0,0 +1,75 @@
+
+    // Formats used for parsing of date and time from user input, locale is explicity set to en_sg for compatibility
+    private static final Locale LOCALE = new Locale("en_sg");
+    protected static final DateTimeFormatter DATETIME_FORMATTER = DateTimeFormatter.ofPattern("ha yyyy/MM/dd",
```
I like how you structure the accessibility of local variables (protected, private)

SamsonChew reviewed (commented):
Overall the code structure is organized, just a few coding standard violations.

NicholasT1 committed:
Add GUI with JavaFX

NicholasT1 committed:
Change Error message for find

NicholasT1 committed:
Merge branch 'Level-10'

NicholasT1 committed:
Add some javadoc comments and align with code style

NicholasT1 committed:
Add asserts for relevant methods

Asserts are not used.

Some errors may be hidden and are not easily identifiable.

Let's use asserts at important points to ease debugging

NicholasT1 committed:
Adjust some Code quality issues

Storage class has a long method.
This hampers readability and does not make full use of abstraction.

Let's extract out methods so that the code is more readable.

NicholasT1 committed:
Merge pull request #2 from NicholasT1/A-Assertions

Add asserts for relevant methods

Asserts are not used.

Some errors may be hidden and are not easily identifiable.

Let's use asserts at important points to ease debugging

NicholasT1 committed:
Merge pull request #3 from NicholasT1/A-CodeQuality

Storage class has a long method.
This hampers readability and does not make full use of abstraction.

Let's extract out methods so that the code is more readable.

NicholasT1 committed:
Add commands to load tasks from a different file

YapBot could only load from a fixed file location which limited its
functionality.

Being able to select which file to load from adds more versatility.

NicholasT1 committed:
Merge branch 'BCD-Extension'

YapBot could only load from a fixed file location which limited its
functionality.

Being able to select which file to load from adds more versatility.

NicholasT1 committed:
Add javadocs to some crucial methods

Some important methods do not have javadocs comments.
This would make the code more difficult to use for others.

Let's add relevant javadoc comments, particularly for important methods.

NicholasT1 committed:
Update test cases and add new test cases

Some test cases are not relevant.
This results in unreliable tests.

Let's update the test cases for more reliable testing.

NicholasT1 committed:
Update readme.md

Readme file does not contain all features and does not serve as a good
user guide.

Let's elaborate on all features of YapBot and add in a screenshot for
better readability.

NicholasT1 committed:
Add slate theme and set title for userguide

Title of user guide contains repository name which hampers user
experience.

Let's set the title to be YapBot and add a theme to improve user
experience
